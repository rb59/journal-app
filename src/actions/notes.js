import Swal from 'sweetalert2';
import { db } from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types/types';
// cd

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
        };
        try {
            const doc = await db
                .collection(`${uid}/journal/notes`)
                .add(newNote);
            dispatch(activeNote(doc.id, newNote));
            dispatch(addNote(doc.id, newNote));
        } catch (error) {
            console.log(error);
        }
    };
};

export const addNote = (id, note) => ({
    type: types.addNew,
    payload: {
        id,
        ...note,
    },
});

export const activeNote = (id, note) => ({
    type: types.setActive,
    payload: {
        id,
        ...note,
    },
});

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        try {
            const notes = await loadNotes(uid);
            dispatch(setNotes(notes));
        } catch (error) {
            console.log(error);
        }
    };
};

export const setNotes = (notes) => ({
    type: types.loadNotes,
    payload: notes,
});

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        if (!note.url) {
            delete note.url;
        }
        note.date = new Date().getTime();
        const noteToFirestore = { ...note };
        delete noteToFirestore.id;
        try {
            await db
                .collection(`${uid}/journal/notes`)
                .doc(note.id)
                .update(noteToFirestore);
            dispatch(refreshNote(note.id, note));
            Swal.fire(
                'Saved',
                `<b><i>"${note.title}"</i></b> has been saved`,
                'success'
            );
        } catch (error) {
            console.log(error);
        }
    };
};

export const refreshNote = (id, note) => ({
    type: types.updateNote,
    payload: { id, note },
});

export const startUploadPicture = (file) => {
    return async (dispatch, getState) => {
        const { active } = getState().notes;
        Swal.fire({
            title: 'Uploading picture...',
            text: 'Please wait...',
            allowOutsideClick: false,
        });
        Swal.showLoading();
        try {
            const fileUrl = await fileUpload(file);
            active.url = fileUrl;
            dispatch(startSaveNote(active));
        } catch (error) {
            console.log(error);
        }
        Swal.close();
    };
};

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        try {
            await db.doc(`${uid}/journal/notes/${id}`).delete();
            dispatch(deleteNote(id));
            Swal.fire('Deleted', 'Your note has been deleted', 'error');
        } catch (error) {
            console.log(error);
        }
    };
};

export const deleteNote = (id) => ({
    type: types.deleteNote,
    payload: id,
});

export const cleanNotes = () => ({
    type: types.logoutCleaning,
});
