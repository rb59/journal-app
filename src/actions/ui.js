import { types } from '../types/types';

export const setError = (err) => ({
    type: types.setError,
    payload: err,
});
export const unsetError = (err) => ({
    type: types.unsetError,
});
export const startLoading = () => ({
    type: types.startLoading,
});
export const finishLoading = () => ({
    type: types.finishLoading,
});
