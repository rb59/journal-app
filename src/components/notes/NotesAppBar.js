import { DateTime } from 'luxon';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
    startDeleting,
    startSaveNote,
    startUploadPicture,
} from '../../actions/notes';
// cd
export const NotesAppBar = () => {
    const { active } = useSelector((state) => state.notes);
    const dispatch = useDispatch();
    const date = DateTime.now().toLocaleString(DateTime.DATE_FULL);
    const handleAddPicture = () => {
        document.querySelector('#fileSelector').click();
    };
    const handleSave = () => {
        dispatch(startSaveNote(active));
    };
    const handleDelete = () => {
        dispatch(startDeleting(active.id));
    };
    const handleFileChange = (e) => {
        const file = e.target.files[0];
        if (file) {
            dispatch(startUploadPicture(file));
        }
    };

    return (
        <div className="notes__appbar">
            <span>{date}</span>
            <input
                id="fileSelector"
                name="file"
                type="file"
                style={{ display: 'none' }}
                onChange={handleFileChange}
            />
            <div>
                <button className="btn" onClick={handleAddPicture}>
                    Picture
                </button>
                <button className="btn" onClick={handleSave}>
                    Save
                </button>
                <button className="btn btn-danger" onClick={handleDelete}>
                    Delete
                </button>
            </div>
        </div>
    );
};
