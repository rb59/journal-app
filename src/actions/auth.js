import Swal from 'sweetalert2';
import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";

export const startLogin = (email,password) => {
    return async(dispatch) => {
        dispatch( startLoading() );
        try {
            const {user:{uid,displayName}} = await firebase.auth().signInWithEmailAndPassword(email,password);  
            dispatch( login(uid,displayName) ); 
            dispatch( finishLoading() );
        } catch (e) {
            Swal.fire('Error',e.message,'error');
            dispatch( finishLoading() );
        }
    };
};

export const startRegister = (name,email,password) => {
    return async(dispatch) => {
        dispatch( startLoading() );
        try {
            const {user} = await firebase.auth().createUserWithEmailAndPassword(email,password);
            await user.updateProfile({displayName: name});
            const {uid,displayName} = user;
            dispatch( login(uid,displayName) );
            dispatch( finishLoading() );  
        } catch (e) {
            Swal.fire('Error',e.message,'error');
            dispatch( finishLoading() );  
        }
    };
};

export const startGoogleLogin = () => {
    return async(dispatch) => {
        try {
            const {user:{uid,displayName}} = await firebase.auth().signInWithPopup(googleAuthProvider);
            dispatch( login(uid,displayName) );
        } catch (e) {
            console.log(e);
        }
    };
};

export const login = (uid,displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName,
    },
});

export const startLogout = () => {
    return async(dispatch) => {
        await firebase.auth().signOut();
        dispatch( logout() );
    };
}; 

export const logout = () => ({
    type: types.logout,
});