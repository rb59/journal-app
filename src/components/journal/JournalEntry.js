import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { DateTime } from 'luxon';
import { activeNote } from '../../actions/notes';
export const JournalEntry = ({ id, title, body, date, url }) => {
    const { active } = useSelector((state) => state.notes);
    const dispatch = useDispatch();
    const noteDate = DateTime.fromMillis(date).toLocal();
    const handleSelectEntry = () => {
        const note = {
            title,
            body,
            date,
            url,
        };
        dispatch(activeNote(id, note));
    };
    return (
        <div
            className={`journal__entry ${
                active?.id === id ? 'selected' : ''
            } animate__animated animate__fadeIn animate__faster`}
            onClick={handleSelectEntry}
        >
            {url && (
                <div
                    className="journal__entry-picture"
                    style={{
                        backgroundImage: `url(${url})`,
                        backgroundSize: 'cover',
                    }}
                ></div>
            )}
            <div className="journal__entry-body">
                <p className="journal__entry-title">{title}</p>
                <p className="journal__entry-content">{body}</p>
            </div>
            <div className="journal__entry-date-box">
                <span>{noteDate.toFormat('cccc')}</span>
                <h4>{noteDate.toFormat('dd')}</h4>
            </div>
        </div>
    );
};
