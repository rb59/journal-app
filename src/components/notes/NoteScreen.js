import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';
// cd
export const NoteScreen = () => {
    const dispatch = useDispatch();
    const { active } = useSelector((state) => state.notes);
    const [values, handleInputChange, reset] = useForm(active);
    const { title, body } = values;
    const activeId = useRef(active.id);

    useEffect(() => {
        if (activeId.current !== active.id) {
            reset(active);
            activeId.current = active.id;
        }
    }, [active, reset]);

    useEffect(() => {
        dispatch(activeNote(values.id, { ...values }));
    }, [values, dispatch]);

    return (
        <div className="notes__main-content">
            <NotesAppBar />
            <div className="notes__content">
                <input
                    type="text"
                    name="title"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    value={title}
                    onChange={handleInputChange}
                />
                <textarea
                    name="body"
                    placeholder="What happened today"
                    className="notes__textarea"
                    value={body}
                    onChange={handleInputChange}
                ></textarea>
                {active.url && (
                    <div className="notes__image mt-1">
                        <img alt="note" src={active.url} />
                    </div>
                )}
            </div>
        </div>
    );
};
