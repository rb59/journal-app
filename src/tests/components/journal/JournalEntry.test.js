import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { activeNote } from '../../../actions/notes';
import { JournalEntry } from '../../../components/journal/JournalEntry';
// cd
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initStore = {
    notes: {
        active: {
            id: '123aer',
            title: 'titulo',
            body: 'cuerpo',
            date: 1627335006001,
        },
    },
};

let store = mockStore(initStore);
store.dispatch = jest.fn();

describe('Pruebas en JournalEntry', () => {
    beforeEach(() => {
        store.clearActions();
        jest.clearAllMocks();
    });
    const wrapper = mount(
        <Provider store={store}>
            <JournalEntry {...initStore.notes.active} />
        </Provider>
    );

    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe disparar el activeNote', () => {
        wrapper.find('.journal__entry').simulate('click');

        expect(store.dispatch).toHaveBeenCalledWith(
            activeNote(initStore.notes.active.id, { ...initStore.notes.active })
        );
    });
});
