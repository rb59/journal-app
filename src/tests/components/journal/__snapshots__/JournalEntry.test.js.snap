// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en JournalEntry debe mostrarse correctamente 1`] = `
<div
  className="journal__entry selected animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      titulo
    </p>
    <p
      className="journal__entry-content"
    >
      cuerpo
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Monday
    </span>
    <h4>
      26
    </h4>
  </div>
</div>
`;
